include/kernels/copy_field_offset.cuh

```diff
@@ -0,0 +1,251 @@
+
+#include <comm_key.h>
+
+namespace quda
+{
+
+  template <class Field_, class Element_, class Accessor_, QudaPCType pc_type_ = QUDA_4D_PC> struct CopyFieldOffsetArg {
+
+    static constexpr int nDim = 4; // No matter what the underlying field is, the dimension is 4
+    static constexpr QudaPCType pc_type = pc_type_;
+
+    using Field = Field_;
+    using Element = Element_;
+    using Accessor = Accessor_;
+
+    Accessor out;      // output field
+    const Accessor in; // input field
+
+    int_fastdiv X0h_in;
+    int_fastdiv X0h_out;
+    int_fastdiv dim_in[nDim];  // full lattice dimensions
+    int_fastdiv dim_out[nDim]; // full lattice dimensions
+
+    CommKey offset;
+
+    int nParity;
+
+    int volume_cb_in;
+    int volume_4d_cb_in;
+
+    int volume_cb_out;
+    int volume_4d_cb_out;
+
+    int volume_cb;
+    int volume_4d_cb;
+
+    int Ls; // The fifth dimension size
+
+    QudaOffsetCopyMode mode;
+
+    CopyFieldOffsetArg(Accessor &out_accessor, Field &out_field, const Accessor &in_accessor, const Field &in_field,
+                       CommKey offset) :
+      out(out_accessor), in(in_accessor), nParity(in_field.SiteSubset()), offset(offset)
+    {
+      const int *X_in = in_field.X();
+      const int *X_out = out_field.X();
+
+      Ls = in_field.Ndim() == 4 ? 1 : X_in[4];
+
+      if (Ls > 1 && X_out[4] != Ls) { errorQuda("Ls mismatch: in: %d, out: %d", X_out[4], Ls); }
+
+      if ((offset[0] + offset[1] + offset[2] + offset[3]) % 2 == 1) {
+        // This offset would change the parity between the input and output fields.
+        errorQuda("Offset (%d,%d,%d,%d) not supported", offset[0], offset[1], offset[2], offset[3]);
+      }
+
+      for (int d = 0; d < nDim; d++) {
+        dim_out[d] = out_field.full_dim(d);
+        dim_in[d] = in_field.full_dim(d);
+      }
+
+      X0h_out = dim_out[0] / 2;
+      X0h_in = dim_in[0] / 2;
+
+      volume_cb_in = in_field.VolumeCB();
+      volume_cb_out = out_field.VolumeCB();
+
+      if (volume_cb_out > volume_cb_in) {
+        volume_cb = volume_cb_in;
+        mode = QudaOffsetCopyMode::COLLECT;
+      } else {
+        volume_cb = volume_cb_out;
+        mode = QudaOffsetCopyMode::DISPERSE;
+      }
+
+      volume_4d_cb_in = volume_cb_in / Ls;
+      volume_4d_cb_out = volume_cb_out / Ls;
+      volume_4d_cb = volume_cb / Ls;
+    }
+  };
+
+  template <class Arg>
+  __device__ __host__ inline
+    typename std::enable_if<std::is_same<typename Arg::Field, ColorSpinorField>::value, void>::type
+    copy_field(int out, int in, int parity, Arg &arg)
+  {
+    using Element = typename Arg::Element;
+
+    Element element = arg.in(in, parity);
+    arg.out(out, parity) = element;
+  }
+
+  template <class Arg>
+  __device__ __host__ inline typename std::enable_if<std::is_same<typename Arg::Field, CloverField>::value, void>::type
+  copy_field(int out, int in, int parity, Arg &arg)
+  {
+    using Element = typename Arg::Element;
+    constexpr int length = 72;
```

Comment:
Is there some way to query this from `CloverField` or an accessor thereof as opposed to a "magic" 72?

Reply:
The magic 72 is something that's been around for a long time and is prevalent in a bunch of clover code already.  It's something that I'll clean up at some point, and I wouldn't burden @hummingtree with this here 😉